tidy3d/plugins/mode/mode_solver.py

```diff
@@ -514,17 +521,27 @@ def _solve_all_freqs(
         symmetry: Tuple[Symmetry, Symmetry],
     ) -> Tuple[List[float], List[Dict[str, ArrayComplex4D]], List[EpsSpecType]]:
         """Call the mode solver at all requested frequencies."""
+        
+        # Get the number of cpu cores
+        n_cpus = os.cpu_count()
 
-        fields = []
-        n_complex = []
-        eps_spec = []
-        for freq in self.freqs:
-            n_freq, fields_freq, eps_spec_freq = self._solve_single_freq(
-                freq=freq, coords=coords, symmetry=symmetry
-            )
-            fields.append(fields_freq)
-            n_complex.append(n_freq)
-            eps_spec.append(eps_spec_freq)
+        if n_cpus <= 2: 
+            n_jobs = 1 # just use one worker if the number of cpus is less than or equal to 2
+        else:
+            # using n_jobs=-2 will result in all CPUs but one being used. n_jobs= -1 will results in all CPUs being use.
+            n_jobs = DEFAULT_NUM_JOB 
+
+        # Use multiple python worker processes to speed up when solving for mutiple frequency points
+        results = Parallel(n_jobs=n_jobs)(delayed(self._solve_single_freq)(
+                                                                            freq = freq, 
+                                                                            coords = coords, 
+                                                                            symmetry = symmetry) 
+                                                                            for freq in self.freqs
+                                                                            )
```

Comment:
Interestingly we do something very similar on the backend, which is why we have this `_solve_single_freq` in the first place. We use `multiprocessing` there. I'm even wondering if we shouldn't just be bringing backend code here. But yeah the explicit thread setting is generally required or else things can get substantially slower. Similarly one reason I didn't want to put this on the frontend is that this can get stuck or crash if the user runs out of memory (big mode solves in prallel). Do you think this is a problem in your implementation?

Reply:
It does seem like `joblib` provides a resource to avoid the over-subscription, but it doesn't seem like you are using it?

https://joblib.readthedocs.io/en/latest/parallel.html#avoiding-over-subscription-of-cpu-resources

Reply:
@momchil-flex So, what I am doing is I am looking at the number of cores the local computer has. If the number of cores is greater than 2, I use all but 1 core for parallel computation. So, if I have 4 cores, the code uses all 3 cores. 
If the backend code has `multiprocessing` I would assume the same implementation also works on the frontend. I have tested this with 100s of frequency simulations but have not tested this is big mode solvers in parallel. Do you have some examples I can use to test this? I can also just make a really fine mesh and see how it does?

Reply:
There are two things here. The first one is the memory. I think you should definitely test with some large mode planes. `multiprocessing` does not handle that smoothly, it can definitely freeze if you run out of memory. So on the backend we do some checks to estimate how much memory will be needed to decide how many cores to use.

The other point is about disabling multi-threading *within each process* to avoid over-allocation which can completely slow things down. This is what the link to joblib that I posted is about. I remember you experienced this in your initial testing (you needed to set the env variables I sent you to 1), but here you don't seem to be doing that. So I'm wondering what's happening, whether it's joblib doing something (which it seems to be able to, but you're not using that functionality as far as I can tell), or something else.

Reply:
@momchil-flex okay, that makes sense to set `inner_max_num_threads`=1. Apparently, by default for "the 'loky' backend: by default each worker process will have environment variables set to allow a maximum of cpu_count() // n_jobs (which is 1 in my case with 12 CPUs and (12-1) njobs )so that the total number of threads used by all the workers does not exceed the number of CPUs of the host." 

This is probably why it was working fine. 

 So, now I have something like this written down and it works the same as before

`        with parallel_config(backend="loky", inner_max_num_threads=1):
            results = Parallel(n_jobs=n_jobs)(delayed(self._solve_single_freq)(
                                                                                freq = freq, 
                                                                                coords = coords, 
                                                                                symmetry = symmetry) 
                                                                                for freq in self.freqs
                                                                                )
                                                                                `

Reply:
@momchil-flex okay tried it with crazy fine with resolution of `min_steps_per_wvl= 1000` compared to 15 before (so around 4400 times larger number of grid points). So, this time it crashed saying system out of memory. We need to implement similar checks like you had in the backend then. Could we discuss this tomorrow?